projects/dsp-ui/src/lib/core/session.service.ts

```diff
@@ -0,0 +1,215 @@
+}
+
+/**
+ * Session with id (= login timestamp) and inforamtion about logged-in user
+ */
+export interface Session {
+    id: number;
+    user: CurrentUser;
+}
+
+@Injectable({
+    providedIn: 'root'
+})
+export class SessionService {
+
+    /**
+     * max session time in milliseconds
+     * default value (24h): 86400000
+     *
+     */
+    readonly MAX_SESSION_TIME: number = 86400000; // 1d = 24 * 60 * 60 * 1000
+
+
+    constructor(
+        @Inject(DspApiConnectionToken) private dspApiConnection: KnoraApiConnection,
+        @Inject(DspApiConfigToken) private dspApiConfig: KnoraApiConfig
+    ) { }
+
+    /**
+     * get session information from localstorage
+     */
+    getSession(): Session {
+        if (localStorage.getItem('session') !== null) {
+            return JSON.parse(localStorage.getItem('session'));
+        }
+
+        return undefined;
+    }
+
+    /**
+     * set session by using the json web token (jwt) and the user object;
+     * it will be used in the login process
+     *
+     * @param jwt
+     * @param username
+     */
+    setSession(jwt: string, identifier: string, identifierType: 'email' | 'username') {
+
+        let session: Session;
+
+        if (jwt) {
+            this.updateDspApiConnection(jwt);
+        }
+
+        // get user information
+        this.dspApiConnection.admin.usersEndpoint.getUser(identifierType, identifier).subscribe(
+            (response: ApiResponseData<UserResponse>) => {
+                let sysAdmin = false;
+                const projectAdmin: string[] = [];
+
+                // get permission information: a) is user sysadmin? b) get list of project iri's where user is project admin
+                const groupsPerProjectKeys: string[] = Object.keys(response.body.user.permissions.groupsPerProject);
+
+                for (const key of groupsPerProjectKeys) {
+                    if (key === Constants.SystemProjectIRI) {
+                        sysAdmin = response.body.user.permissions.groupsPerProject[key].indexOf(Constants.SystemAdminGroupIRI) > -1;
+                    }
+
+                    if (response.body.user.permissions.groupsPerProject[key].indexOf(Constants.ProjectAdminGroupIRI) > -1) {
+                        projectAdmin.push(key);
+                    }
+                }
+
+                // store session information in browser's localstorage
+                // TODO: jwt will be removed, when we have a better cookie solution (DSP-261)
+                session = {
+                    id: this.setTimestamp(),
+                    user: {
+                        name: response.body.user.username,
+                        jwt: jwt,
+                        lang: response.body.user.lang,
+                        sysAdmin: sysAdmin,
+                        projectAdmin: projectAdmin
+                    }
+                };
+
+                // update localStorage
+                localStorage.setItem('session', JSON.stringify(session));
```

Comment:
this is only executed once the answer from Knora arrives. So after calling `setSession` there is no way to know when this is done.

Could this be be done in a better way? For example, the method could return an `Observable` but there could be a `pipe` that handles writing the information to the cache.

Reply:
with RxJS, you can perform actions without subscribing: https://www.learnrxjs.io/learn-rxjs/operators/utility/do

The subscriber would be the one that calls `setSession`. Once the subscriber gets notified, the session has been set.

Reply:
Resolved in c3eb394